### ISSUE TITLE
open-vm-tools 11.1.0 I/O error when copying files from Fedora VM to Windows Host even with workaround

### ISSUE THREAD

**Issue by hermidalc** [vmware/open-vm-tools#437]
Am I doing the open-vm-tools 11.1.0 workaround wrong?  I've always had enable shared folders by default and here is my fstab entry, which I've also always had and has worked until upgrading to 11.1.0:
```bash
.host:/  /mnt/hgfs  fuse.vmhgfs-fuse  allow_other,uid=1000,gid=1000,defaults  0 0
```
The mounts are there but when I try to copy something only in the direction from the Fedora 32 VM to the Windows host I get I/O errors.  Any thoughts on how to get this working??

**Comment by NoelJB** [comment]
HGFS works for me, but after a reboot I find that I have to disable and enable them before it all works.  But once I do, HGFS works (until the next guest reboot).

**Comment by oliverkurth** [comment]
You mean you disable/enable sharing in Workstation? That is a good hint - maybe it's trying to mount too early.

**Comment by hermidalc** [comment]
> HGFS works for me, but after a reboot I find that I have to disable and enable them before it all works. But once I do, HGFS works (until the next guest reboot).

But the problem I'm having is that I've always had an fstab entry for vmhgfs-fuse, and the 11.1.0 [release notes](https://github.com/vmware/open-vm-tools/blob/stable-11.1.x/ReleaseNotes.md) state that this will workaround the issue permanently.

My mounts are there in the Linux VM, so unlike the release notes issue where it says they will be unavailable I can see them and do ls, etc.  But when I try to copy something from Linux VM to Windows I get I/O error.

If I downgrade back to 11.0.5 the problem is gone.

**Comment by hermidalc** [comment]
> You mean you disable/enable sharing in Workstation? That is a good hint - maybe it's trying to mount too early.

Having the fstab entry does not totally fix the problem.  You still cannot copy files from Linux VM to Windows.

**Comment by oliverkurth** [comment]
Please try the steps outlined here: https://kb.vmware.com/s/article/74650 . It recommends to use systemd instead, and wait for fuse to be ready.

**Comment by oliverkurth** [comment]
Another question - was this working for you before upgrading to 11.1.0? If so, which version was working? I want to make sure if this is a regression.

**Comment by oliverkurth** [comment]
@NoelJB , your issue seems to be different. The advise to use the systemd approach as outlined in the kb article  https://kb.vmware.com/s/article/74650 may fix your issue. If not, please create a separate issue on github.

I filed an internal bug directed at the hgfs team for the I/O that @hermidalc observed.

**Comment by hermidalc** [comment]
> 
> 
> Another question - was this working for you before upgrading to 11.1.0? If so, which version was working? I want to make sure if this is a regression.

@oliverkurth, yes everything works again when you downgrade back to 11.0.5.  I still have to try your recommended solution and will report back here.

**Comment by hermidalc** [comment]
> @NoelJB , your issue seems to be different. The advise to use the systemd approach as outlined in the kb article https://kb.vmware.com/s/article/74650 may fix your issue. If not, please create a separate issue on github.
> 
> I filed an internal bug directed at the hgfs team for the I/O errors that @hermidalc observed.

@oliverkurth the https://kb.vmware.com/s/article/74650 solution works and tested with 11.1.0 thank you!  So I believe the older method of using fstab (which also creates a systemd unit) doesn't support enough configuration like you said to make sure it mounts after fuse is ready, so the direct systemd config is now the preferred approach.

@NoelJB I would recommend trying the solution above and not using fstab.  You should never have to worry about needing to check/uncheck things after every guest reboot (I never did).

**Comment by hermidalc** [comment]
Hi @oliverkurth - I'm not sure if some recent update in Fedora 32 changed things since I last wrote, but now I'm getting a similar but sligtly different issue, even with the new systemd fuse and vmhgfs config.  It seems like if I haven't used fuse in the VM for a certain period of time, I get an `Input/output error` when doing an operation.  If I immediately run the same command again it then works, so it's like it deactivated or something if you haven't used it for a while?

**Comment by atkdef** [comment]
Not sure if relevant but I got something similar. 

The guest OS is Debian Bullseye, did apt upgrade yesterday and found out that the shared directory is not working right now. 
It seems that from the guest, I can only copy small files (<64KB) into the shared directory. 
If the file is larger than that, it says "input/output error". 

Neither the fstab method nor the systemd method mentioned above worked. 

Any thoughts? I'm willing to do further tests / provide more info if needed.

**Comment by zazen-jp** [comment]
Fedora 32 VM in Fusion. Exactly the same symptoms as atkdef. New systemd fuse and vmhgfs config does not help
Everything fine with 11.0.5.

**Comment by hermidalc** [comment]
> Hi @oliverkurth - I'm not sure if some recent update in Fedora 32 changed things since I last wrote, but now I'm getting a similar but sligtly different issue, even with the new systemd fuse and vmhgfs config. It seems like if I haven't used fuse in the VM for a certain period of time, I get an `Input/output error` when doing an operation. If I immediately run the same command again it then works, so it's like it deactivated or something if you haven't used it for a while?

@oliverkurth @thomasvmw I found the source of the above issue, if I suspend the VM and then unsuspend it then I  get the `Input/output error` once.  Then it works after the first failed command.  But if I suspend and unsuspend it again, it happens once again.  VM reboots do not have this problem.  This also didn't occur in 11.0.5

**Comment by bburman** [comment]
I am experiencing the same issue as @atkdef and @zazen-jp. My guest OS is the latest Parrot OS. I also didn't have the issue until after I upgraded to open-vm-tools 11.1. Switching from fstab to systemd did not solve my problem. I can only copy small files to these mount points. Rebooting the VM does not work.

**Comment by Bacto** [comment]
Hi,
I have the same issue. I can copy very small files but biggest trigger an "I/O error".
I'm using Alpine Linux 3.12 as guest and VMware Fusion.

```
# vmhgfs-fuse --version
vmhgfs-fuse: version 1.6.12.0

FUSE library version: 2.9.9
fusermount version: 2.9.9
using FUSE kernel interface version 7.19
```

```
# vmtoolsd --version
VMware Tools daemon, version 11.1.0.19887 (build-16036546)
```

**Comment by CalculonPrime** [comment]
I'm also a victim of the bug in v11.1.0.  Upgraded from OpenSUSE leap 15.1 => 15.2 yesterday which upgraded to v11.1.0.  I just discovered that I get an input/output error when trying to copy a large file to a share.  The workaround from https://kb.vmware.com/s/article/74650 doesn't work for me.

I manually downloaded the older version rpms for v11.0.5 from the Leap 15.1 repos and installed them manually and now everything is working again.  I protected the three v11.0.5 libraries in yast so they're not blown away during my next update, but now I'm asked to do a conflict resolution every time I update (who thought that was proper behavior)?

How could this have happened?  If there aren't "copy a large file to a share" autotests being run before releasing this project, they should be.  I really had better uses for my Sunday afternoon.

**Comment by horshack-dpreview** [comment]
Same issue for me. After some trial and error I determined the error only occurs when the I/O size for the copy is >= 64KB, which the GNOME UI shell uses, along with apps such as Firefox (downloading files into the share) and shell utilities such as cp. The only way I've been able to work around it is by using dd and specifying a block size less than 64K, such as:

dd if='/home/user/Pictures/screenshot.png' of=/mnt/hgfs/myshare/screenshot.png bs=63k

If I specify 64k in the above command line I get a generic "Input/output error".

I don't see the error when copying in the other direction, ie from my Windows host share to my Ubuntu guest from within the Ubuntu guest.

I'm running Ubuntu 20.04, open-vm-tools 2:11.1.0-2~ubuntu20.04.1 on the guest VM and Windows 10 on the host. I see the error for every share I've tried to copy to.

**Comment by gnbond** [comment]
> You can work around the issue by adding a max_write=61440 to the mount options in fstab (64K - 4K). For example:
> 
> .host:/ /mnt/hgfs fuse.vmhgfs-fuse noauto,allow_other,max_write=61440 0 0

I can confirm this fixed it for me after I had the same EIO issue on large files after I upgraded the guest OS from ubuntu 18 (open-vm-tools 11.0.5) to ubuntu 20 (open-vm-tools 11.1.0)

Thanks a heap, @horshack-dpreview !

**Comment by fatso83** [comment]
Also started getting error after upgrading to Ubuntu 20.04 (package version `2:11.1.0-2~ubuntu20.04.1`). Adding the `max_write` option fixed the issue without a hitch!

**Comment by Bacto** [comment]
Any news from VMware team for this bug that is opened since 4 months now?

**Comment by denji** [comment]
Any solution without dirty hacks?

**Comment by ravindravmw** [comment]
I hope someone will respond on this by next week. I have requested the concerned developers to look into it.

**Comment by CalculonPrime** [comment]
Will this ever get fixed in an official release?  We're approaching 5 months since the bug was reported.  There have been two releases but it's not been fixed.  Pretty shocking.  I don't want to have to install hacks.

**Comment by jinwvmw** [comment]
We cannot internally reproduce the file copying error when I/O size >=64KB, but since @horshack-dpreview has provided a workaround, we will have further investigation based on this workaround and will update the status soon.

**Comment by yliuvm** [comment]
I setup a totally new environment about win10 host and ubuntu 2010/2004 guest, and can't reproduce the file copying error when size >= 64KB.
I wonder if there was special configurations or something causing this problem.
Could anyone provide more details?

@NoelJB ，I can reproduce the issue you met that need to disable and enable hgfs to make it works after reboot, and I'm working on it.

**Comment by horshack-dpreview** [comment]
> I setup a totally new environment about win10 host and ubuntu 2010/2004 guest, and can't reproduce the file copying error when size >= 64KB.
> I wonder if there was special configurations or something causing this problem.
> Could anyone provide more details?
> 
> @NoelJB ，I can reproduce the issue you met that need to disable and enable hgfs to make it works after reboot, and I'm working on it.

Just to make sure we're on the same page, here are the quick commands to reproduce:

First create a 64KB data file on / inside the Ubuntu VM:
`dd if=/dev/zero of=test64k.bin bs=64KB count=1`

Then try to copy the file from / to a share on the Windows 10 host, using a 64KB I/O size:
`dd if=test64k.bin of=/mnt/hgfs/myshare/test64k.bin bs=64k count=1`

This fails on my 20.04 LTS system if I don't have max_write set to < 64KB on the fuse entry in my fstab.

**Comment by yliuvm** [comment]
@horshack-dpreview, I tried your way, but still not reproduce the error.
My guest OS:
Ubuntu 20.04 LTS
GNOME shell 3.36.4

I'm not sure it's os version related issue, there are kinds of guest os mentioned in the comments.
We tried most of them, but can not reproduce the error.
I will invest it based on the workaround, and any comments is appreciated.

**Comment by horshack-dpreview** [comment]
> @horshack-dpreview, I tried your way, but still not reproduce the error.
> My guest OS:
> Ubuntu 20.04 LTS
> GNOME shell 3.36.4
> 
> I'm not sure it's os version related issue, there are kinds of guest os mentioned in the comments.
> We tried most of them, but can not reproduce the error.
> I will invest it based on the workaround, and any comments is appreciated.

It fails on my VMware Pro Workstation version 15.5.1 build-15018445. I checked for updates and saw 15.5.7 build-17171714 was available. I just installed that and it no longer fails.

**Comment by horrordash** [comment]
I confirm. Ubuntu 20.04 LTS, open-vm-tools 11.1.5. write error occurs on VMware Player 15.5.1. it works fine on 15.5.7

**Comment by andlabs** [comment]
Let me try, though my situation is different still:

VMware Fusion 11.5.3 (15870345); I can't upgrade because I'm staying on macOS 10.12 here. vmhgfs-fuse version is from Ubuntu:

```
vmhgfs-fuse: version 1.6.12.0

FUSE library version: 2.9.9
fusermount version: 2.9.9
using FUSE kernel interface version 7.19
```

This is after a `do-release-upgrade` from 19.10 to 20.10. I do NOT mount my shares on bootup or in fstab, but instead invoke `vmware-hgfs` manually:

```
# (run from $HOME)
/usr/bin/vmhgfs-fuse .host:/vmshare vmshare -o subtype=vmhgfs-fuse
```

I get the Input/output error if I `cp` a large file in, or if I use shell redirection. If I instead pipe my file's data into `tee` and have `tee` write the file instead, it works perfectly. And yes, the file size does have to be >= 64KB — in fact, the limit for a successful copy appears to be 65458 bytes:

```
pietro@ubuntu:~/games$ dd if=/dev/zero of=x bs=1 count=65457
65457+0 records in
65457+0 records out
65457 bytes (65 kB, 64 KiB) copied, 0.215802 s, 303 kB/s
pietro@ubuntu:~/games$ cp x ~/vmshare/test
pietro@ubuntu:~/games$ dd if=/dev/zero of=x bs=1 count=65458
65458+0 records in
65458+0 records out
65458 bytes (65 kB, 64 KiB) copied, 0.220569 s, 297 kB/s
pietro@ubuntu:~/games$ cp x ~/vmshare/test
cp: error writing '/home/pietro/vmshare/test': Input/output error
```

Using `dd` directly also seems to work, even with the default output block size of 512 bytes. I wonder if that's why `tee` also works.

As shown, it does not matter if the file exists already or not.

Copying from the shared folder into the VM works fine, and changes to the shared folder from outside the VM are reflected properly.

Hopefully this helps!

**Comment by gkyildirim** [comment]
> Let me try, though my situation is different still:
> 
> VMware Fusion 11.5.3 (15870345); I can't upgrade because I'm staying on macOS 10.12 here. vmhgfs-fuse version is from Ubuntu:
> 
> ```
> vmhgfs-fuse: version 1.6.12.0
> 
> FUSE library version: 2.9.9
> fusermount version: 2.9.9
> using FUSE kernel interface version 7.19
> ```
> 
> This is after a `do-release-upgrade` from 19.10 to 20.10. I do NOT mount my shares on bootup or in fstab, but instead invoke `vmware-hgfs` manually:
> 
> ```
> # (run from $HOME)
> /usr/bin/vmhgfs-fuse .host:/vmshare vmshare -o subtype=vmhgfs-fuse
> ```
> 
> I get the Input/output error if I `cp` a large file from the VM out to the shared folder, or if I use shell redirection. If I instead pipe my file's data into `tee` and have `tee` write the file instead, it works perfectly. And yes, the file size does have to be >= 64KB — in fact, the limit for a successful copy appears to be 65458 bytes:
> 
> ```
> pietro@ubuntu:~/games$ dd if=/dev/zero of=x bs=1 count=65457
> 65457+0 records in
> 65457+0 records out
> 65457 bytes (65 kB, 64 KiB) copied, 0.215802 s, 303 kB/s
> pietro@ubuntu:~/games$ cp x ~/vmshare/test
> pietro@ubuntu:~/games$ dd if=/dev/zero of=x bs=1 count=65458
> 65458+0 records in
> 65458+0 records out
> 65458 bytes (65 kB, 64 KiB) copied, 0.220569 s, 297 kB/s
> pietro@ubuntu:~/games$ cp x ~/vmshare/test
> cp: error writing '/home/pietro/vmshare/test': Input/output error
> ```
> 
> Using `dd` directly also seems to work, even with the default output block size of 512 bytes. I wonder if that's why `tee` also works.
> 
> As shown, it does not matter if the file exists already or not.
> 
> Copying from the shared folder into the VM works fine, and changes to the shared folder from outside the VM are reflected properly.
> 
> Hopefully this helps!

I had a similiar issue and it's solved by adding `max_write=32768` to my fstab.
This is my fstab:
`.host:/storage-vm	/root/mac	fuse.vmhgfs-fuse	auto,allow_other,max_write=32768	0	0`

I am using **Ubuntu** 20.04.3 LTS VM, **Macos** 10.13.6 and **Vmware Fusion** 11.5.3. 
And also I have the Ubuntu latest **openvmtools** (open-vm-tools/focal-updates,now 2:11.3.0-2ubuntu0~ubuntu20.04.1) which is same as my quoted comment.
```
vmhgfs-fuse: version 1.6.12.0

FUSE library version: 2.9.9
fusermount version: 2.9.9
using FUSE kernel interface version 7.19
```

**Comment by andlabs** [comment]
Late but yes, I can confirm the max_write workaround works, even with the current Ubuntu/open-vm-tools 1.6.12.0. Thanks!

**Comment by DaniloMV** [comment]
> Am I doing the open-vm-tools 11.1.0 workaround wrong? I've always had enable shared folders by default and here is my fstab entry, which I've also always had and has worked until upgrading to 11.1.0:
> 
> ```shell
> .host:/  /mnt/hgfs  fuse.vmhgfs-fuse  allow_other,uid=1000,gid=1000,defaults  0 0
> ```
> 
> The mounts are there but when I try to copy something only in the direction from the Fedora 32 VM to the Windows host I get I/O errors. Any thoughts on how to get this working??

It worked on Windows 11 main OS with a VMware 16 Player where I have Ubuntu 20.04, it shared the files perfectly. Thanks for sharing hermidalc

**Comment by hermidalc** [comment]
> > @NoelJB , your issue seems to be different. The advise to use the systemd approach as outlined in the kb article https://kb.vmware.com/s/article/74650 may fix your issue. If not, please create a separate issue on github.
> > I filed an internal bug directed at the hgfs team for the I/O errors that @hermidalc observed.
> 
> @oliverkurth the https://kb.vmware.com/s/article/74650 solution works and tested with 11.1.0 thank you! So I believe the older method of using fstab (which also creates a systemd unit) doesn't support enough configuration like you said to make sure it mounts after fuse is ready, so the direct systemd config is now the preferred approach.
> 
> @NoelJB I would recommend trying the solution above and not using fstab. You should never have to worry about needing to check/uncheck things after every guest reboot (I never did).

@DaniloMV use the systemd approach, I don’t think many use the fstab approach anymore I certainly haven’t in a very long time

**Comment by hermidalc** [comment]
> > > @NoelJB , your issue seems to be different. The advise to use the systemd approach as outlined in the kb article https://kb.vmware.com/s/article/74650 may fix your issue. If not, please create a separate issue on github.
> > > I filed an internal bug directed at the hgfs team for the I/O errors that @hermidalc observed.
> > 
> > 
> > @oliverkurth the https://kb.vmware.com/s/article/74650 solution works and tested with 11.1.0 thank you! So I believe the older method of using fstab (which also creates a systemd unit) doesn't support enough configuration like you said to make sure it mounts after fuse is ready, so the direct systemd config is now the preferred approach.
> > @NoelJB I would recommend trying the solution above and not using fstab. You should never have to worry about needing to check/uncheck things after every guest reboot (I never did).
> 
> @DaniloMV use the systemd approach, I don’t think many use the fstab approach anymore I certainly haven’t in a very long time

And to others as well looking here, only changes to the VMware KB instructions I made are in the `Options` I added:

```
Options=default_permissions,allow_other,uid=1000,gid=1000,auto_unmount
```

Otherwise it is a root mount and not a my own user.

**Comment by DaniloMV** [comment]
> > > > @NoelJB , your issue seems to be different. The advise to use the systemd approach as outlined in the kb article https://kb.vmware.com/s/article/74650 may fix your issue. If not, please create a separate issue on github.
> > > > I filed an internal bug directed at the hgfs team for the I/O errors that @hermidalc observed.
> > > 
> > > 
> > > @oliverkurth the https://kb.vmware.com/s/article/74650 solution works and tested with 11.1.0 thank you! So I believe the older method of using fstab (which also creates a systemd unit) doesn't support enough configuration like you said to make sure it mounts after fuse is ready, so the direct systemd config is now the preferred approach.
> > > @NoelJB I would recommend trying the solution above and not using fstab. You should never have to worry about needing to check/uncheck things after every guest reboot (I never did).
> > 
> > 
> > @DaniloMV use the systemd approach, I don’t think many use the fstab approach anymore I certainly haven’t in a very long time
> 
> And to others as well looking here, only changes to the VMware KB instructions I made are in the `Options` I added:
> 
> ```
> Options=default_permissions,allow_other,uid=1000,gid=1000,auto_unmount
> ```
> 
> Otherwise it is a root mount and not a my own user.

@hermidalc I understand I will take you into account and use systemd.
Thanks for the explanation.

**Comment by hermidalc** [comment]
This issue doesn't occur anymore in recent versions of FUSE, hgfs, and open-vm-tools. If someone is still experiencing it I will reopen the issue

**Comment by cappallo** [comment]
I had this problem and like others, it was solved by adding max_write=32768. But I recently updated to Ubuntu LTS 22.04 and it seems like it uses fuse3 which no longer supports max_write option, so now this problem is back; I can't copy any files out to a share which are of any substantial size.

**Comment by johnwvmw** [comment]
@cappallo   As you have noticed, Linux vendors are replacing fuse (fuse-2) with fuse3 and fuse 3 does not have the "max_write" option.

I have had no problems copying 5 MB to 500+ MB files to and from the "shared folder" on UB 22.05 with open-vm-tools 11.3.5.  That may not meet your definition of "substantial", but I am working on a guest with limited disk space.

I think that if you remove the "max_write=nnnn" option from the other options on the "vmhgfs-fuse" line in your "/etc/fstab" file, transfers will work.   

Since this issue was file against an earlier open-vm-tools release and fuse-2, and we are discussing fuse 3 and a later OVT release, please use a "new" issue for any follow-up. 

Incidentally, Ubuntu has made open-vm-tools 12.1.0 available for UB 22.04 LTS.    You may want to upgrade version if you have not already done so.
